keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,655 @@
+
+No.
+
+### Troubleshooting
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+The job controller will declare that the job is "Succeeded" based on the `.status.completedIndexes`.
+So, in this case, the job controller can not correctly evaluate the successPolicy 
+because the `.status.completedIndexes` won't be updated.
+
+###### What are other known failure modes?
+
+None.
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+If a successPolicy isn't respected even though the job doesn't match other policies 
+such as a [pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/) and backoffLimit:
+
+- Check reachability between Kubernetes components.
+- Consider increasing the logging level to trace when the issues occur.
+- Check the job controller's `job_sync_duration_seconds` metric to check if the job controller's processing duration increases. 
+
+If many requests are rejected, re-queued many times or increased the job controller's processing duration,
+consider tuning the parameters for [APF](https://kubernetes.io/docs/concepts/cluster-administration/flow-control/).
+
+## Implementation History
+ 
+- 2023.06.06: This KEP is ready for review.
+- 2023.06.09: API design is updated.
+- 2023.10.03: API design is updated.
+
+## Drawbacks
+
+Adds more complexity to the criteria to be terminated Job.
+
+## Alternatives
+
+### Relax a validation for completions
+Currently, a job is restricted `.spec.completion!=nil`.
+By relaxing the validation, the job can be declared as succeeded when some indexes succeeded,
+similar to NonIndexed jobs.
+
+### Hold succeededIndexes as []int typed in successPolicy
+
+We can hold the `succeededIndexes` as []int typed that a job can be declared as succeeded.
+
+```golang
+// SuccessPolicyOnSucceededPods describes a Job can be succeeded based on succeeded pods.
+type SuccessPolicyTarget struct {
+	// Specifies a set of required indexes.
+	// The job is declared as succeeded if a set of indexes are succeeded. 
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// 
+	// +optional
+	SucceededIndexes []int
+...
+}
+```
+
+However, if we allow users to set all `succeededIndexes` (`0-10^5`) to `.spec.successPolicy.target.succeededIndexes`
+and don't limit the number of list sizes, there are cases that the size of a job object is too big.
+
+In the worst case, allowed all indexes (`0-10^5`) are added to `succeededIndexes`, and a `succeededIndexes` size
+is `SUM[9*10^n*(n+1)]+2+6≈5.6656MiB`, where:
+- `n` starts from `0` and goes up to `5`.
+- `1` of `n+1` means a separator that indicates `,`.
+- `2` is the sum of the index `0` and `,`.
+- `6` is the size of indexes `10^5`.
+
+So, if we select this alternative API design, we need to limit the size of `indexes`. 
+
```

Comment:
How about adding this:

> ### Match indexes using CEL
> 
> Use CEL to define which indexes are matched. For example:
>
> ```yaml
> ...
> spec:
>   parallelism: 7
>   completions: 7
>   successPolicy:
>     - target:
>         succeededCount: 1
>         indexMatchExpression: "index == (job.completions - 1)" # always the last index
>         whenTargetAchieved: terminate
> ```
>
> or
> 
> ```yaml
> ...
> spec:
>   parallelism: 42
 >  completions: 42
>   successPolicy:
>     - target:
>         succeededCount: 21 # exactly half
>         indexMatchExpression: "(index % 2) == 1" # odd numbers
>         whenTargetAchieved: continue
> ```

(imagine that the even numbered Pods do one kind of work that is valuable, but not essential, and the odd numbered Pods produce the actual result that I need)


---

Defining an alternative doesn't mean selecting it, but it does let us explain why we ruled it out.

Reply:
That's a good alternative. I will add this way here as a separate alternative.
Thanks.

Reply:
Done.